### PR TITLE
User profile

### PR DESCRIPTION
**Pull request by Tramchandani**
Screenshot from 2021-03-22 15-47-55
Screenshot from 2021-03-17 14-21-40
Screenshot from 2021-03-17 14-21-36
Screenshot from 2021-03-17 14-21-31

### CODE REVIEW

**Comment by kalpana19sharma** [comment]
@Tramchandani please resolve the conflicts.

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -4,22 +4,66 @@ def new
...
+      flash[:notice] = "signup successful"
+      redirect_to user_path(@user.id)
     else
       render :new
     end
   end
 
   def show
-    @user = User.find_by_id(session[:user_id])   
+    @user = current_user
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: Use a before action filter to set `@user`

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -4,22 +4,66 @@ def new
...
+    @user = current_user  
+  end
+
+  def edit
+    @user = current_user
+  end
+
+  def update
+    @user = current_user
+    flash[:notice] = "updating profile"
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: What's the need for this?

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -4,22 +4,66 @@ def new
...
+    end
+  end
+
+  def edit_password
+    @user = current_user
+  end
+
+  def update_password
+    @user = current_user
+    if @user&.authenticate(password_params[:current_password]) && @user.update(password: password_params[:password], password_confirmation: password_params[:password_confirmation])
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: Line is too long.

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (kalpana19sharma) [reply app/controllers/users_controller.rb]: This line of code is repetitive

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -7,6 +7,8 @@
     <%= csp_meta_tag %>
 
     <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= stylesheet_link_tag "https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" %>
+    <%= stylesheet_link_tag "https://cdnjs.cloudflare.com/ajax/libs/material-design-iconic-font/2.2.0/css/material-design-iconic-font.min.css" %>
```
Comment (kalpana19sharma) [inline app/views/layouts/application.html.erb]: Don't use the cdn url for bootstrap. We are going to integrate package for bootstrap.

File: app/views/users/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    <%= form.label :email %>
+    <%= form.email_field :email %>
+    <p class = "error"><%= show_errors(@user, :email) %></p>
+    
+    <%= form.submit "Update", class: 'btn btn-primary btn-block' %>
+  <% end %>
+  <%= link_to "back", user_path(session[:user_id]) %>
+  <%= render partial: "layouts/show_notice" %>
+  <%= render partial: "layouts/display_error" %>
+</div>
```
Comment (kalpana19sharma) [inline app/views/users/edit.html.erb]: add new line

File: app/views/users/edit_password.html.erb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<%= render partial: "layouts/sidebar" %>
+<h1> Edit password </h1>
+<div class = "form-container">
+  <%= form_for :user_pass , url: :update_password do |form| %>
```
Comment (kalpana19sharma) [inline app/views/users/edit_password.html.erb]: is it correct? `url: :update_password`

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (Tramchandani) [reply app/controllers/users_controller.rb]: ma'am I added it for debugging purpose.
I will remove it now

File: app/views/layouts/application.html.erb (hunk 4 above)
Comment (Tramchandani) [reply app/views/layouts/application.html.erb]: ma'am how do we integrate specific urls through bootstrap package?

Comment (kalpana19sharma) [reply app/views/layouts/application.html.erb]: yes

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -1,25 +1,67 @@
 class UsersController < ApplicationController
+  before_action :current_user
+  skip_before_action :current_user, only: [:new, :create]
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: Can we combine these 2 lines.

File: app/models/user.rb [hunk 8]
```diff
@@ -1,7 +1,11 @@
 class User < ApplicationRecord
   REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
   has_secure_password
-  validates :first_name, :email, :password, presence: true
-  validates :password, length: { in: 3..20 }, confirmation: true
+  validates :first_name, :email, presence: true
+  validates :password, presence: true, length: { in: 3..20 }, confirmation: true, if: :updating_profile?
   validates :email, uniqueness: { case_sensitive: false }, format: { with: REGEX }
+
+  def updating_profile?
```
Comment (kalpana19sharma) [inline app/models/user.rb]: Rename this method to `should_validate_password?`

File: app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -7,6 +7,7 @@
     <%= csp_meta_tag %>
 
     <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= stylesheet_link_tag "https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" %>
```
Comment (kalpana19sharma) [inline app/views/layouts/application.html.erb]: Remove this URL in bootstrap PR

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/models/user.rb (hunk 8 above)
Comment (kalpana19sharma) [reply app/models/user.rb]: Also what should be the access level of this method?

File: app/views/layouts/application.html.erb (hunk 9 above)
Comment (Tramchandani) [reply app/views/layouts/application.html.erb]: ma'am I have tried using bootstrap classes and they are working fine but yet if I am removing this link the UI gets distorted.
I tried different things to fix it by integrating bootstrap again in different ways but still it's not working properly.

Comment (Tramchandani) [reply app/views/layouts/application.html.erb]: I even tried adding a feature from bootstrap 4 documentation and it is working fine but the url is for version 3.3.7 so maybe the code I have used uses some feature of earlier versions which is not supported by bootstrap 4 that I have integrated in the rails app

Comment (kalpana19sharma) [reply app/views/layouts/application.html.erb]: I think this link is including bootstrap 3. If this can be the issue?

File: app/models/user.rb [hunk 10]
```diff
@@ -1,7 +1,13 @@
...
-  validates :password, length: { in: 3..20 }, confirmation: true
+  validates :first_name, :email, presence: true
+  validates :password, presence: true, length: { in: 3..20 }, confirmation: true, if: :should_validate_password?
   validates :email, uniqueness: { case_sensitive: false }, format: { with: REGEX }
+
+	private
+
+  def should_validate_password?
+    new_record? || password.present?
+  end
```
Comment (kalpana19sharma) [inline app/models/user.rb]: Fix indentation

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/views/layouts/application.html.erb (hunk 9 above)
Comment (Tramchandani) [reply app/views/layouts/application.html.erb]: Yes I think that's the issue. Maybe bootstrap 4 does not support some of the classes of bootstrap 3 and I might have used them in my code. So is there a way to add support for bootstrap 3 as well as 4 in the application?

Comment (kalpana19sharma) [reply app/views/layouts/application.html.erb]: I don't think so. Let's move to bootstrap 4

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -1,25 +1,68 @@
...
+      flash[:notice] = "signup successful"
+      redirect_to user_path(@user.id)
     else
       render :new
     end
   end
 
   def show
-    @user = User.find_by_id(session[:user_id])   
+    render layout: 'empty'
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb:20]: Why do we need a different layout?

File: app/views/layouts/_sidebar.html.erb [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        <%= link_to edit_user_path(session[:user_id]) do%>
+          <i class = "zmdi zmdi-border-color"> </i> 
+          Edit Profile
+        <% end %>
+      </li>
+      <li>
+        <%= link_to edit_password_path(session[:user_id]) do%>
+          <strong><i class="zmdi zmdi-lock"></strong></i> 
+          Edit Password
+        <% end %>
```
Comment (kalpana19sharma) [inline app/views/layouts/_sidebar.html.erb]: Use proper spaces and check all occurrences

File: app/views/layouts/_sidebar.html.erb [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+      <li>
+        <%= link_to logout_path do%>
+          <i class = "zmdi zmdi-long-arrow-return"> </i>
+          Logout
+        <% end %>
+      </li>
+    </ul>
+  </div>
+</div>
+
```
Comment (kalpana19sharma) [inline app/views/layouts/_sidebar.html.erb]: Remove extra line

File: app/views/users/edit.html.erb [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
```
Comment (kalpana19sharma) [inline app/views/users/edit.html.erb]: Remove extra line

File: app/views/users/show.html.erb [hunk 15]
```diff
@@ -1,3 +1,61 @@
...
+              </span>
+            </div>
+            <%= render partial: "layouts/show_notice" %>
+          </div>
+        </div>
+      </div>
+    </div>
+  </div>
+</div>
+
```
Comment (kalpana19sharma) [inline app/views/users/show.html.erb]: Remove extra line

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (Tramchandani) [reply app/controllers/users_controller.rb:20]: In application.html.erb I have center aligned all the elements.
But to display user profile I had to display info in a larger area hence the different layout.

Comment (kalpana19sharma) [reply app/controllers/users_controller.rb:20]: ok

File: app/views/layouts/_sidebar.html.erb (hunk 12 above)
Comment (Tramchandani) [reply app/views/layouts/_sidebar.html.erb]: all occurances as in?

**Review by kalpana19sharma — APPROVED** [review]

**Pull request merged**